### ISSUE TITLE
ema-bin-??? and possible preproc failure

### ISSUE THREAD

**Issue by michaeljmetzger** [arshajii/ema#38]
I am not sure if ema-bin-??? refers to something I should be replacing. When I try to run the ema align step I get the error:

error: file /home/metzgerm/EMA/ema-bin-??? could not be opened

The preproc does not make any files or folders with "ema-bin" in the name. Should it? The preproc log is short and unclear:

:: Bucketing         1 inputs into     500 files with       40 threads
:: Loading known counts file counts_file.ema-ncnt ... 
:: Loading known counts ... done in 4.9 s
:: Loading full counts file counts_file.ema-fcnt ... 
::: Loading dump      à?        aê‘1 of size   @ 14,913,081 (   Bec          910 MB) done in 2.052 s

I have run:
cat ~/EMA/some_clam_S3_L004_interleaved.fastq | ema count -w ~/EMA/4M-with-alts-february-2016.txt -o counts_file 2>&1 | tee count.log
cat ~/EMA/some_clam_S3_L004_interleaved.fastq | ema preproc -w ~/EMA/4M-with-alts-february-2016.txt -n 500 -t 40 -o ~/EMA *.ema-ncnt 2>&1 | tee preproc.log
#parallel --bar -j10 "ema align -t 4 -d -r ~/arrow/Mar.3.3.3.consensus.fasta -s {} | samtools sort -@ 4 -O bam -l 0 -m 4G -o {}.bam -" ::: ~/EMA/ema-bin-???

Thanks for your help.

Best,
Michael

**Comment by arshajii** [comment]
Hi,
Yes, `ema preproc` should produce 500 (`-n`) files of the form `ema-bin-000`, `ema-bin-001` etc. (the `???` is a bash idiom for matching files of this form). I wonder if there is some quirk with that FASTQ file -- could you post the first 10 or so lines of it?

**Comment by arshajii** [comment]
That FASTQ looks good -- how do the `*.ema-ncnt` files look? Are they produced by `ema count`?

**Comment by michaeljmetzger** [comment]
No clue how the .ema-ncnt files really look. They are not text. I have only two files: counts_file.ema-ncnt and counts_file.ema-fcnt. The following is the first 10 lines of counts_file.ema-ncnt:
�	*
9�
-�
�
��

��
�
�
�c

The log for the count is:
:: Loading 10X took 2.4 s
:: Dumped block           1
:: Dumped block           2
:: Dumped block           3
:: Dumped block           4
:: Dumped block           5
:: Dumped block           6
:: Dumped block           7
:: Dumped block          8
:: Counting took 2472.2 s
:: Reads with OK barcode: Ò637,148,268 out of  Ð683,891,843
:: Ignored th OK barcode: Ò637,1480 reads
:: Dumped block          9
:: Printing took 8.6 s
:: Processed oo683,891,843 reads (          470,201 MB uncompressed) in   2,483 s

--Thanks for your help!

**Comment by inumanag** [comment]
Hi @michaeljmetzger 

Can you please try doing this instead:

```
cat ~/EMA/some_clam_S3_L004_interleaved.fastq | ema preproc -w ~/EMA/4M-with-alts-february-2016.txt -n 500 -t 40 -o test *.ema-ncnt 2>&1 | tee preproc.log
```

Is`test` directory empty or not? I have a feeling that passing `-o ~/EMA` confuses the script because of `~`.

**Comment by michaeljmetzger** [comment]
Unfortunately, that still doesn’t generate any files other than the log file. I tried creating an empty directory called test and tried again without any directory called test in the working directory. I also tried putting all files in the working directory so there is no ~ anywhere in the command, but no change. The log is always some variation of:

:: Bucketing ________1 inputs into ____500 files with ______40 threads
:: Loading known counts file counts_file.ema-ncnt ...
:: Loading known counts ... done in 3.9 s
:: Loading full counts file counts_file.ema-fcnt ...
::: Loading dump ø=>åd_À________±Õ_1 of size __@_14,913,081 (__ »Šª___ _______910 MB) done in 1.859 s

Thanks so much for your help on this.

--Michael

**Comment by inumanag** [comment]
So the log contains all these non-ASCII characters? Weird... any chance you can share the first 10k records from the interleaved file so that I can take a look?

**Comment by michaeljmetzger** [comment]
Yes, even the ema count log has a few odd non ASCII characters that make it confusing to interpret:
cat ~/EMA/some_clam_S3_L004_interleaved.fastq | ema count -w ~/EMA/4M-with-alts-february-2016.txt -o counts_file 2>&1 | tee count.log

:: Loading 10X took 2.4 s
:: Dumped block __________1
:: Dumped block __________2
:: Dumped block __________3
:: Dumped block __________4
:: Dumped block __________5
:: Dumped block __________6
:: Dumped block __________7
:: Dumped block _________8
:: Counting took 2472.2 s
:: Reads with OK barcode: _Ò637,148,268 out of _Ð683,891,843
:: Ignored th OK barcode: _Ò637,1480 reads
:: Dumped block _________9
:: Printing took 8.6 s
:: Processed oo683,891,843 reads (__________470,201 MB uncompressed) in __2,483 s

I assumed this was just a quirk of how it runs, but this could suggest something fundamentally wrong here.

I can make a file with the first 10k reads and try to upload that if you think that could help troubleshooting.

Do you know of a small test dataset I could use to test the configuration of the program?

Thanks!
--Michael

**Comment by inumanag** [comment]
OK, that is certainly unexpected.

Which OS are you using? You can email me the data link (`inumanag at mit dot edu`--- first 10k reads should be fine) and I can take a look. Currently, we don't have any small sample for these purposes.

**Comment by michaeljmetzger** [comment]
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)

running python 2.7 in this module

I will send a file with the first 10k pairs.

Thanks!
--Michael

**Comment by inumanag** [comment]
Hi @michaeljmetzger 

Just following up on this one: did you manage to resolve this issue with another compiler or pre-compiled binaries?

**Comment by michaeljmetzger** [comment]
Sorry, I haven’t had a chance to try reinstalling it yet on our servers. Longranger align appears to have run successfully for us with this dataset, however. Thanks for all your help on this. Hopefully I can give the reinstallation a shot in the future to compare the EMA mapping to long ranger align mapping. Would you have any reason to think they would be different?
Best,
Michael

Michael J. Metzger
Assistant Investigator
Pacific Northwest Research Institute
720 Broadway
Seattle, WA 98122
metzgerm@pnri.org
metzgerm@uw.edu
206-330-1977
http://www.pnri.org/research-programs/metzger-lab/
________________________________

**Comment by inumanag** [comment]
Hi @michaeljmetzger 

Yeah, I believe that there is something wrong with the compiler on your cluster because the generated binary is completely corrupted--- probably because some compiler library versions got mixed up (this happens sometimes on cluster environments). 

As for Long Ranger vs. EMA: yeah, they are. We found out that EMA's results are more robust, esp. in repetitive regions, so we always map with EMA.

**Comment by michaeljmetzger** [comment]
I have now given it one more shot and installed using conda. I can see the help text when I type "ema help" but the command yields a completely empty log file and no count files at all and no errors shown.

cat ~/EMA/some_clam_S3_L004_interleaved.fastq | ema count -w ~/EMA/4M-with-alts-february-2016.txt -o counts_file 2>&1 | tee count.log

Any thoughts?
